README.md
This is a framework for communication between your php website and the roomsXML service API.

Hint: Read the roomsXML API specification / FAQ before using this framework. You need
to know what a region id is, what a booking prepare request is and so on.


#### \#0 How to use:
- create an instance of the RoomsXML class. Use the methods to set your options.
    you can/must specify urls, input date format, running mode (live/dev) and many more (see [\# 1](#1-rooms-xml-options)).
    This is probably only necessary once per execution of your program. You stick to this roomsXML object
    all the time.

- once you have set up your RoomsXML object, you can almost start to interact with the service.
    Follow the instructions below every time you want to perform the request.

- create an instance of the Vacation class. here you specify all information regarding your query (see [\# 2](#2-how-to-start-a-vacation)).
    you can/must specify number of adults, number of kids, start and end date, etc.

- now perform the request towards roomsXML using the methods provided by the RoomsXML class (see [\# 3](#3-roomsxml-requests)).
     These are for example: availability_search_region, availability_search_hotel, booking_prepare,...
     
#### \#1 Rooms XML options

<table>
  <tr>
    <th>Option</th>
    <th>Explanation</th>
    <th>Method</th>
    <th>Parameters</th>
  </tr>
  <tr>
    <td>Running Mode</td>
    <td>If in development mode, <br>the framework will use the dev URLs,<br>Live mode will use the live URLs.</td>
    <td>set_running_mode($mode)</td>
    <td>String: 'dev' or 'live'</td>
  </tr>
  <tr>
    <td>Language</td>
    <td>The language you want to speak <br>with RoomsXML. <br>Accepts an ISO language identifier</td>
    <td>set_language($language)</td>
    <td>String:  'en', 'fr', etc.</td>
  </tr>
  <tr>
    <td>Api Version</td>
    <td>The version of the RoomsXML API</td>
    <td>set_api_ver($ver)</td>
    <td>String: e.g. '1.25'</td>
  </tr>
  <tr>
    <td>Input Date Format</td>
    <td>The date format you want to use<br>when you set up a new vacation.<br>Default is 'mm/dd/yyyy'</td>
    <td>set_input_date_format($format)</td>
    <td>String:<br>Allowed delimiters:<br>'-' and '/'.<br>Day: 'dd', Month: 'mm'<br>Year: 'yyyy'. 'yy' doesnt<br>work.</td>
  </tr>
  <tr>
    <td>Currency</td>
    <td>The currency for the return prices.<br>Accepts an ISO-4217 / SWIFT <br>three character currency code<br>(e.g. USD, GBP,EUR).</td>
    <td>set_currency($currency)</td>
    <td>String: e.g. 'EUR'</td>
  </tr>
  <tr>
    <td>Live URL</td>
    <td>The Live URL</td>
    <td>set_live_url($url)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Dev URL</td>
    <td>The Dev URL</td>
    <td>set_dev_url($url)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Organisation</td>
    <td>The organisation name for the<br>log-in at RoomsXML</td>
    <td>set_org($org)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>User</td>
    <td>Your username for log-in</td>
    <td>set_user($user)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Password</td>
    <td>Your Password</td>
    <td>set_password($pw)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Soap Action Availability<br>Search URL</td>
    <td>The soap URL for availability<br>search</td>
    <td>set_soap_action_avail_url($url)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Soap Action Booking URL</td>
    <td>The soap URL for booking</td>
    <td>set_soap_action_book_url($url)</td>
    <td>String</td>
  </tr>
  <tr>
    <td>Soap Action Cancel URL</td>
    <td>The soap URL for cancelling a<br>booking.</td>
    <td>set_soap_action_bookcancel($url)</td>
    <td>String</td>
  </tr>
</table>

#### \#2 How to start a vacation

Parameters for instantiation of a vacation object:

**Please be aware that this framework is not capable of performing multi-room
requests with different configurations
(e.g. 1st room with 2 adults, 2. room with 1 adult and 1 child)

If you specify a vacation with 2 rooms, 1 adult and 2 kids, that 
means you want two rooms with 1 adult and 2 kids each.**

<table>
  <tr>
    <th>Parameter</th>
    <th>Explanation</th>
    <th>Type</th>
    <th>Example<br></th>
  </tr>
  <tr>
    <td>$rooms</td>
    <td>Number of rooms<br></td>
    <td>int</td>
    <td>1<br></td>
  </tr>
  <tr>
    <td>$adults</td>
    <td>Number of adults<br></td>
    <td>int</td>
    <td>2<br></td>
  </tr>
  <tr>
    <td>$kids</td>
    <td>Number of kids<br></td>
    <td>int</td>
    <td>0<br></td>
  </tr>
  <tr>
    <td>$startdate</td>
    <td>Startdate in your<br>specified input format<br></td>
    <td>String</td>
    <td>'05/23/2018'<br></td>
  </tr>
  <tr>
    <td>$enddate</td>
    <td>Enddate in your <br>specified input format<br></td>
    <td>String</td>
    <td>'05/30/2018'</td>
  </tr>
  <tr>
    <td>$child_ages (optional)<br></td>
    <td>Ages of the children<br></td>
    <td>Array</td>
    <td>array(4,5)</td>
  </tr>
  <tr>
    <td>$adult_names<br>(somewhat optional)</td>
    <td>The names of the adults<br>on the vacation.<br>An array of arrays.<br>Each contains title, <br>first and last name<br>(in this order).<br>HINT: for booking<br>requests you need<br>to give a vacation<br>with at least one<br>adult name.<br></td>
    <td>Multi-dim <br>Array</td>
    <td>array(array('Mr.','Ron','Weasley'),<br>array('Mr.','Harry','Potter'))<br></td>
  </tr>
  <tr>
    <td>$child_names(optional)</td>
    <td>Child Names.<br>Data structure like <br>adult names above.<br></td>
    <td>Multi-dim<br>Array</td>
    <td>array(array('Ms.'Childy','Child')<br></td>
  </tr>
  <tr>
    <td>$minprice(optional)</td>
    <td>Minimum price in<br>specified currency.<br></td>
    <td>int</td>
    <td>200<br></td>
  </tr>
  <tr>
    <td>$maxprice(optional)</td>
    <td>Maximum price in<br>specified currency<br></td>
    <td>int</td>
    <td>1000</td>
  </tr>
</table>

#### \#3 RoomsXML requests

Finally you want to send your requests t0 the
roomsXML service API!

You can do this using the methods below.

<table>
  <tr>
    <th>Method</th>
    <th>Explanation</th>
    <th>Parameters</th>
  </tr>
  <tr>
    <td>availability_search_hotel($hotel_id, $vacation, $return_type)</td>
    <td>Availability search for a specific hotel by ID.<br>If the return type is 'array' you get an array<br>with all relevant hotel information.<br>If the return type is 'obj' you get an<br>object that also contains the requests and<br>the responses as string. </td>
    <td>$hotel_id: int<br>$vacation: Vacation<br>$return_type: String<br> 'array'/'obj'</td>
  </tr>
  <tr>
    <td>availability_search_region($region_id,$vacation,$return_type)</td>
    <td>Like above but search for all hotels in a<br>specific region. You get a list with all region IDs<br>from roomsXML.</td>
    <td>$region_id: int<br>$vacation: Vacation<br>$return_type: String<br>'array'/'obj'</td>
  </tr>
  <tr>
    <td>booking_prepare($quote_id,$vacation,$return_type)</td>
    <td>Performs a booking request in prepare <br>mode with the quote_id from an availability search.<br>You need to do this twice before you can do <br>an actual booking at roomsXML. Return<br>type like availability_search_hotel.</td>
    <td>$quote_id: int<br>$vacation: Vacation<br>$return_type: String<br>'array'/'obj'</td>
  </tr>
  <tr>
    <td>booking_confirm($quote_id, $vacation, $return_type)</td>
    <td>Performs a real booking request with a quote_id<br>from an availability search. Note that 1. you have<br>to run booking prepare twice before and<br>2. you are not allowed to give a vacation without<br>at least one adult name!</td>
    <td>$quote_id: int<br>$vacation: Vacation<br>$return_type: String<br>'array'/'obj'</td>
  </tr>
  <tr>
    <td>booking_cancel($booking_id)</td>
    <td>Cancels the booking with the given booking id.</td>
    <td>$booking_id: int</td>
  </tr>
</table>







changelog
2018-02-01

- vacation object includes nationality, this will be used in the availability search.
- if no (or not enough) guest names are set, generate unique guest names
- getter for RoomType, MealType and AdditionalInformation
- some changes in variable naming

test/tester.php
<?php

/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 15/09/17
 * Time: 16:10
 */

function loadFrameworkClass($name) {
	require_once( __DIR__ . "/../roomsXML/" . $name . ".php");
}

spl_autoload_register("loadFrameworkClass");

$r = new RoomsXML();
$r->set_currency( 'EUR' );
$r->set_dev_url("http://www.roomsxmldemo.com/RXLStagingServices/ASMX/XmlService.asmx");
$r->set_running_mode( RoomsXML::$mode_dev );
$r->set_org( 'sps' );
$r->set_user( 'xmltest' );
$r->set_password( 'xmltest' );
$r->set_api_ver( '1.25' );
$r->set_language( 'en' );
$r->set_soap_action_avail_url("http://www.reservwire.com/namespace/WebServices/Xml/AvailabilitySearch");
$r->set_soap_action_book_url("http://www.reservwire.com/namespace/WebServices/Xml/BookingCreate");
$r->set_soap_action_bookcancel("http://www.reservwire.com/namespace/WebServices/Xml/BookingCancel");

$urlaub=new Vacation(1,1,2,'11/01/2019', '11/06/2019',array(12,13));
$obj = $r->availability_search_region( 52612, $urlaub, RoomsXML::$return_type_obj );

$ids=$obj->get_result_ids() ;
print_r( $obj->get_result_ids() );
$quote=0;
if ($ids) {
	$obj2 = $r->booking_prepare( $ids[0], $urlaub, RoomsXML::$return_type_obj );
	$attempts=0;
	foreach ($ids as $id) {
		echo "attempts: $attempts";
		$attempts++;
		$obj2 = $r->booking_prepare( $id, $urlaub, RoomsXML::$return_type_obj );
		if ( ! $obj2->has_failed() ) {
			$quote = $id;
			break;
		}
	}
	echo "booking prepare successful";
	$obj3=$r->booking_confirm( $quote, $urlaub, $obj );
	print_r( $obj3 );
}





test/canceltester.php
<?php
/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 18/09/17
 * Time: 15:20
 */
function loadFrameworkClass($name) {
	require_once( __DIR__ . "/../roomsXML/" . $name . ".php");
}

spl_autoload_register("loadFrameworkClass");
$r = new RoomsXML();
$r->set_currency( 'EUR' );
$r->set_dev_url("http://www.roomsxmldemo.com/RXLStagingServices/ASMX/XmlService.asmx");
$r->set_running_mode( RoomsXML::$mode_dev );
$r->set_org( 'sps' );
$r->set_user( 'xmltest' );
$r->set_password( 'xmltest' );
$r->set_api_ver( '1.25' );
$r->set_language( 'en' );
$r->set_soap_action_avail_url("http://www.reservwire.com/namespace/WebServices/Xml/AvailabilitySearch");
$r->set_soap_action_book_url("http://www.reservwire.com/namespace/WebServices/Xml/BookingCreate");
$r->set_soap_action_bookcancel("http://www.reservwire.com/namespace/WebServices/Xml/BookingCancel");
$res=$r->booking_cancel(720107509);
print_r( $res );

roomsXML/RoomsXML.php
<?php
/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 13/09/17
 * Time: 16:45
 */

//this is the user interface for the roomsxml framework

class RoomsXML {
	public $constants;
	public $api;
	public static $return_type_array='array';
	public static $return_type_obj = 'obj';
	public static $mode_dev='dev';
	public static $mode_live='live';

	function set_running_mode($mode) {
		$this->constants->running_mode=$mode;
	}

	function set_language($language) {
		$this->constants->language=$language;
	}
	function set_api_ver($ver) {
		$this->constants->api_ver = $ver;
	}

	function set_input_date_format($format) {
		$this->constants->input_date_format = $format;
	}

	function __construct() {
		$this->constants=new Constants();
		$this->api=new API($this->constants);
	}

	function set_currency($currency) {
		$this->constants->currency = $currency;
	}

	function set_live_url($url) {
		$this->constants->url_api_live=$url;
	}
	function set_dev_url($url) {
		$this->constants->url_api_dev=$url;
	}
	function set_org($org) {
		$this->constants->org_id = $org;
	}
	function set_user($user) {
		$this->constants->user = $user;
	}
	function set_password($pw) {
		$this->constants->pw = $pw;
	}

	function set_soap_action_avail_url($url) {
		$this->constants->soap_action_avail = $url;
	}

	function set_soap_action_book_url($url) {
		$this->constants->soap_action_book = $url;

	}

	function set_soap_action_bookcancel($url) {
		$this->constants->soap_action_bookcancel = $url;
	}

	function availability_search_hotel($hotel_id, $vacation, $return_type) {

		$xml=$this->api->getRoomsXmlSection($vacation);
		$request=$this->api->getRequestXml($xml, $hotel_id, $vacation, true);
		$pair = RequestResponsePairAvail::create_pair_from_request( $request, $this->api );
		if ($return_type==RoomsXML::$return_type_array) {
			return $pair->get_roomsdata();
		} else {
			return $pair;
		}
	}

	/* @param $vacation Vacation
	 * @param $region_id int

	 */
	function availability_search_region($region_id, $vacation, $return_type) {

		$xml=$this->api->getRoomsXmlSection($vacation);

		$request=$this->api->getRequestXml($xml, $region_id, $vacation, false);
		$pair = RequestResponsePairAvail::create_pair_from_request( $request, $this->api);
		if ($return_type==RoomsXML::$return_type_array) {
			return $pair->get_roomsdata();
		} else {
			return $pair;
		}

	}

	function booking_prepare($quote_id, $vacation, $return_type) {
		$xml=$this->api->getRoomsXmlSection($vacation);
		$request=$this->api->getBookingXML($xml, $quote_id, true);
		$pair = RequestResponsePairBookingCreate::create_pair_from_request( $request, $this->api );
		if ($return_type=='Array') {
			return $pair->get_roomsdata();
		} else {
			return $pair;
		}
	}

	function booking_confirm($quote_id, $vacation, $return_type) {
		$xml=$this->api->getRoomsXmlSection($vacation);
		$request=$this->api->getBookingXML($xml, $quote_id, false);
		$pair = RequestResponsePairBookingCreate::create_pair_from_request( $request, $this->api );
		if ($return_type=='Array') {
			return $pair->get_roomsdata();
		} else {
			return $pair;
		}
	}

	function booking_cancel($booking_id) {
		$request = $this->api->getBookingXmlCancel($booking_id );
		$pair = RequestResponsePairBookingCancel::create_pair_from_request( $request, $this->api );
		return $pair;

	}


}

roomsXML/Vacation.php
<?php
/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 13/09/17
 * Time: 17:04
 */

class Vacation {

	public $rooms;
	public $adults;
	public $kids;
	public $startdate;
	public $enddate;

	/**
	 * @return string
	 */
	public function getStartdate() {
		return $this->startdate;
	}

	/**
	 * @return string
	 */
	public function getEnddate() {
		return $this->enddate;
	}
	public $nationality;
	public $child_ages;
	public $adult_names;
	public $child_names;
	public $minprice;
	public $maxprice;



	function __construct($rooms, $adults, $kids, $startdate, $enddate, $nationality, $child_ages=array(), $adult_names=array(), $child_names=array(), $minprice=0, $maxprice=null) {
		$this->rooms = $rooms;
		$this->adults = $adults;
		$this->kids = $kids;
		$this->startdate = $startdate;
		$this->enddate = $enddate;
		$this->child_ages=$child_ages;
		$this->adult_names=$adult_names;
		$this->child_names=$child_names;
		$this->minprice=$minprice;
		$this->maxprice = $maxprice==null ? API::$no_max_price : $maxprice;
		$this->nationality=$nationality;
	}


}

roomsXML/RequestResponsePairAvail.php
<?php

/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 7/5/17
 * Time: 6:53 PM
 */
class RequestResponsePairAvail extends RequestResponsePair {
	public static $key = 'request_response_pair_avail';
	public static $current_pair;

	private $result_ids;
	private $acc_ids;
	private $roomsdata;

	/**
	 * @return mixed
	 */
	public function get_roomsdata() {
		return $this->roomsdata;
	}


	public function get_result_ids() {
		return $this->result_ids;
	}
	public function get_acc_ids() {
		return $this->acc_ids;
	}



	private function get_result_ids_from_roomsdata() {
		$result_ids=array();
		foreach ($this->roomsdata as $availability_entry) {
			foreach ($availability_entry['Result'] as $result_entry) {
				$result_ids[]=$result_entry['@attributes']['id'];
			}
		}
		return $result_ids;

	}

	private function get_acc_ids_from_roomsdata() {
		$acc_ids=array();
		foreach ( $this->roomsdata as $availability_entry ) {
			$acc_ids[] = $availability_entry['Hotel']['@attributes']['id'];
		}
		return $acc_ids;
	}

	/**
	 * RequestResponsePairAvail constructor.
	 *
	 * @param $request
	 * @param $api API
	 */
	private function __construct($request, $api) {
		$this->response = $api->sendCurlRequest( $request, $api->constants->soap_action_avail );
		$this->request = $request;
		$this->roomsdata = $api->getArrayFromXmlResponseAvail( $this->response );
		$this->result_ids = RequestResponsePairAvail::get_result_ids_from_roomsdata();
		$this->acc_ids=RequestResponsePairAvail::get_acc_ids_from_roomsdata();

	}


	public static function create_pair_from_request($request, $api) {

		$current_pair = new RequestResponsePairAvail( $request, $api );
		$_SESSION [ RequestResponsePairAvail::$key ] = $current_pair;

		RequestResponsePairAvail::$current_pair = $current_pair;
		return $current_pair;
	}


	function get_result_from_id( $result_id ) {
		foreach ( $this->roomsdata as $availability_entry) {
			foreach ( $availability_entry['Result'] as $result_entry ) {
				if ( $result_entry['@attributes']['id'] == $result_id ) {
					return $result_entry;

				}
			}

		}
	}

	function get_result_names_and_meals( $result_id ) {
		$result = $this->get_result_from_id( $result_id);
		if ( is_null( $result ) ) {
			return array();
		}
		$names      = [];

		foreach ( $result['Room'] as $room_entry ) { //result has multiple rooms

			$names[]= array('name' => $room_entry['RoomType']['@attributes']['text'],'meal' => $room_entry['MealType']['@attributes']['text']);

		}

		return $names;

	}


	function get_result_price( $room_type_id ) {
		$result = $this->get_result_from_id( $room_type_id );
		if ( is_null( $result ) ) {
			return null;
		}

		$sum = 0;
		foreach ( $result['Room'] as $room_entry ) {

			$sum += $room_entry['Price']['@attributes']['amt'];
		}
		$room_price = $sum;


		return $room_price;

	}


		}



roomsXML/RequestResponsePairBookingCreate.php
<?php

/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 7/12/17
 * Time: 12:45 PM
 */
class RequestResponsePairBookingCreate extends RequestResponsePair {
	public static $key = 'request_response_pair_booking';
	public static $current_pair;
	private $roomsdata;

	/**
	 * @return mixed
	 */
	public function get_roomsdata() {
		return $this->roomsdata;
	}

	/**
	 * RequestResponsePairBookingCreate constructor.
	 *
	 * @param $request string
	 * @param $api API
	 */
	private function __construct($request, $api) {
		$response = $api->sendCurlRequest( $request, $api->constants->soap_action_book );
		$this->request = $request;
		$this->response = $response;
		$this->roomsdata = $api->getArrayFromXmlResponseBooking( $response );
	}

	public function get_MealTypes() {
		$meal_types=array();
		foreach ( $this->roomsdata['Booking']['HotelBooking'] as $hotelbooking_entry ) {
			$meal_types[]=$hotelbooking_entry['Room']['MealType']['@attributes']['text'];
		}
		return $meal_types;
	}
	public function get_RoomTypes() {
		$room_types=array();
		foreach ($this->roomsdata['Booking']['HotelBooking'] as $hotelbooking_entry) {
			$room_types[]=$hotelbooking_entry['Room']['RoomType']['@attributes']['text'];
		}
		return $room_types;
	}

	public function get_add_infos() {
		$add_infos=array();
		foreach ($this->roomsdata['Booking']['HotelBooking'] as $hotelbooking_entry) {
			$room_infos=array();
			if (array_key_exists('Message',$hotelbooking_entry['Room']['Messages'])) {
				foreach ( $hotelbooking_entry['Room']['Messages']['Message'] as $message_entry ) {
					if ( array_key_exists( 'Type', $message_entry ) ) {
						if ( ! empty( $message_entry['Type'] ) and ! ctype_space( $message_entry['Type'] ) ) {
							$type = esc_html( $message_entry['Type'] );
						}
					}
					if ( array_key_exists( 'Text', $message_entry ) ) {
						if ( ! empty( $message_entry['Text'] ) and ! ctype_space( $message_entry['Text'] ) ) {
							$text = esc_html( $message_entry['Text'] );
						}
					}

					$room_infos[] = ( isset( $type ) ? $type : '' ) . ( ( isset( $type ) and isset( $text ) ) ? ': ' : '' ) . ( isset( $text ) ? $text : '' );

				}
			}

			if ( empty( $room_infos ) ) {
				$room_infos[] = '-';
			}
			$imploded = implode( '<br/>', $room_infos );
			$add_infos[] = html_entity_decode($imploded);
		}

		return $add_infos;
	}


	public static function create_pair_from_request( $request, $api) {

		$pair = new RequestResponsePairBookingCreate( $request, $api);
		self::$current_pair = $pair;
		$_SESSION[ self::$key ] = $pair;

		return $pair;

	}



	function get_raw_price() {
		$sum=0;
		foreach ( $this->roomsdata['Booking']['HotelBooking'] as $hotelbooking_entry ) {
			$sum += $hotelbooking_entry['TotalSellingPrice']['@attributes']['amt'];
		}

		return $sum;

	}



	function has_failed() {
		if ( !$this->response ) {
			return true;
		}


		if (strpos($this->response,"<faultstring>")>0) {
			return true;
		}

		return false;
	}
}

roomsXML/API.php
<?php
/**
 * Created by PhpStorm.
 * User: ullika
 * Date: 5/12/17
 * Time: 3:40 PM
 */


class API {
	public static $no_max_price = 'no_max';
	public $constants;

	/**
	 * API constructor.
	 *
	 * @param $constants Constants
	 */
	public function __construct( $constants ) {
		$this->constants = $constants;
	}


	function sendCurlRequest($xml, $soapaction ) {
		$url_dev=$this->constants->url_api_dev;
		$url_live = $this->constants->url_api_live;



		$mode=$this->constants->running_mode;
		// SET REQ URL
		if ( $mode==RoomsXML::$mode_dev ) {
			$url = $this->constants->url_api_dev;
		} elseif ($mode==RoomsXML::$mode_live) {
			$url = $this->constants->url_api_live;
		} else {
			die("Error: please specify a running mode ('dev'/'live') \n");
		}



		// SET SOAP ACTION
		// SET HEADER
		$header = array(
			"Content-type: text/xml;charset=\"utf-8\"",
			"Accept: text/xml",
			"Cache-Control: no-cache",
			"Pragma: no-cache",
			"SOAPAction: \"" . $soapaction . "\"",
			"Content-length: " . strlen( $xml ),
		);

		// DO SOAP REQUEST
		$soap_do = curl_init();
		curl_setopt( $soap_do, CURLOPT_URL, $url );
		curl_setopt( $soap_do, CURLOPT_CONNECTTIMEOUT, 1000 );
		curl_setopt( $soap_do, CURLOPT_TIMEOUT, 1000 );
		curl_setopt( $soap_do, CURLOPT_RETURNTRANSFER, true );
		curl_setopt( $soap_do, CURLOPT_SSL_VERIFYPEER, false );
		curl_setopt( $soap_do, CURLOPT_SSL_VERIFYHOST, false );
		curl_setopt( $soap_do, CURLOPT_POST, true );
		curl_setopt( $soap_do, CURLOPT_POSTFIELDS, $xml );
		curl_setopt( $soap_do, CURLOPT_HTTPHEADER, $header );
		curl_setopt( $soap_do, CURLOPT_ENCODING, 'gzip,deflate' );
		$response = curl_exec( $soap_do );
		if ( $response === false ) {
			$err = 'Curl error: ' . curl_error( $soap_do );
			curl_close( $soap_do );
			print $err;
		} else {
			curl_close( $soap_do );
		}

		return $response;
	}

	/**
	 * @param $vacation Vacation
	 *
	 * @return null|string
	 */
	public function getRoomsXmlSection( $vacation ) {
		$rooms_xml = null;
		for ( $i = 0; $i < $vacation->rooms; $i ++ ) {
			$rooms_xml .= "<Room>
				      <Guests>";

			for ( $j = 0; $j < $vacation->adults; $j ++ ) {
				$adult_names=$vacation->adult_names;

				if ( $i==0 && array_key_exists( $j,$adult_names ) ) {
					$rooms_xml .= '<Adult title="' . $adult_names[ $j ][0] . '" first="' . $adult_names[ $j ][1] . '" last="' . $adult_names[ $j ][2] . '"></Adult>';
				} else {

					$rooms_xml .= '<Adult title="Mrs." first="test' . str_repeat('R',$i).str_repeat('I',$j) . '" last="testadult" />';
				}
			}
			if ( $vacation->kids ) {
				$child_ages = $vacation->child_ages;
				$child_names = $vacation->child_names;
				$k = 0;

				foreach ( $child_ages as $z => $c ) {

					if ($i==0 && array_key_exists( $k,$child_names ) ) {
						$rooms_xml .= '<Child age="' . $c . '" title="' . $child_names[ $k ][0] . '" first="' . $child_names[ $k ][1] . '" last="' . $child_names[ $k ][2] . '"/>';
					} else {
						$rooms_xml .= '<Child age="' . $c . '" title="Ms." first="test' .str_repeat('R',$i). str_repeat('I',$z) . '" last="testchild"/>';

					}

					$k ++;

				}
			}

			$rooms_xml .= "</Guests>
				    </Room>";
		}

		return $rooms_xml;
	}

	/**
	 * @param $rooms_xml string
	 * @param $id
	 * @param $vacation Vacation
	 * @param bool $is_acc_id
	 *
	 * @return string
	 */
	function getRequestXml( $rooms_xml, $id, $vacation, $is_acc_id = false ) {
		$num_days = $this->getNumDays( $vacation->startdate, $vacation->enddate );
		$formatedStart = $this->convertDate( $vacation->startdate );
		$min_price = $vacation->minprice;
		$max_price = $vacation->maxprice;

		$xml = '<?xml version="1.0"?>
					<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
			        <soap:Body>
			        <AvailabilitySearch xmlns="http://www.reservwire.com/namespace/WebServices/Xml">
			         <xiRequest>
					  <Authority>
					    <Org>' . $this->constants->org_id . '</Org>
					    <User>' . $this->constants->user . '</User>
					    <Password>' . $this->constants->pw . '</Password>
					    <Currency>' . strtoupper( $this->constants->currency ) . '</Currency>
					    <Language>' . $this->constants->language . '</Language>
					    <Version>' . $this->constants->api_ver . '</Version>
					  </Authority>';
		if ( ! $is_acc_id ) {
			$xml = $xml . ' <RegionId>' . $id . '</RegionId>';
		} else {
			$xml = $xml . ' <HotelId>' . $id . '</HotelId>';
		}
		$xml = $xml . '
					  <HotelStayDetails>
					    <ArrivalDate>' . $formatedStart . '</ArrivalDate>
					    <Nights>' . $num_days . '</Nights>
					    <Nationality>'.$vacation->nationality.'</Nationality>
					    ' . $rooms_xml . '
					  </HotelStayDetails>
					  <HotelSearchCriteria>
					    <AvailabilityStatus>allocation</AvailabilityStatus>
					    <DetailLevel>basic</DetailLevel>';
		if ( $min_price != 0 ) {
			$xml = $xml . '<MinPrice>' . $min_price . '</MinPrice>';
		}
		if ( $max_price != API::$no_max_price ) {
			$xml = $xml . '<MaxPrice>' . $max_price . '</MaxPrice>';
		}


		$xml = $xml . '</HotelSearchCriteria>
					  </xiRequest>
					</AvailabilitySearch>
				  </soap:Body>
			</soap:Envelope>';

		return $xml;
	}


	/**
	 * @param $roomsxml
	 * @param $result_id
	 * @param $vacation Vacation
	 * @param bool $dry_run
	 *
	 * @return string
	 */
	function getBookingXML( $roomsxml, $result_id, $dry_run = true ) {

		$mode = ( $dry_run ) ? 'prepare' : 'confirm';

		$xml = '<?xml version="1.0"?>

<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <soap:Body>
    <BookingCreate xmlns="http://www.reservwire.com/namespace/WebServices/Xml">
      <xiRequest>
		<Authority>
		<Org>' . $this->constants->org_id . '</Org>
		<User>' . $this->constants->user . '</User>
		<Password>' . $this->constants->pw . '</Password>
		<Currency>' . strtoupper( $this->constants->currency ) . '</Currency>
		<Language>' . $this->constants->language . '</Language>
		<Version>' . $this->constants->api_ver . '</Version>
		</Authority>
        <QuoteId>' . $result_id . '</QuoteId>
        <HotelStayDetails>
          <ArrivalDate>' . '0001-01-01' . '</ArrivalDate>
          <Nights>' . 0 . '</Nights>
          ' . $roomsxml . '
        </HotelStayDetails>
        <CommitLevel>' . $mode . '</CommitLevel>
      </xiRequest>
    </BookingCreate>
  </soap:Body>
</soap:Envelope>';


		return $xml;
	}


	function getNumDays( $startdate, $enddate ) {

		$dStart = new DateTime( $this->convertDate( $startdate ) );
		$dEnd   = new DateTime( $this->convertDate( $enddate ) );

		$dDiff    = $dStart->diff( $dEnd );
		$num_days = $dDiff->days;

		return $num_days;

	}

	function getBookingXmlCancel( $bookingId ) {
		$xml = '<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
				  <soap:Body>
					<BookingCancel xmlns="http://www.reservwire.com/namespace/WebServices/Xml">
					  <xiRequest>
						<Authority>
						  	<Org>' . $this->constants->org_id . '</Org>
							<User>' . $this->constants->user . '</User>
							<Password>' . $this->constants->pw . '</Password>
							<Currency>' . strtoupper( $this->constants->currency ) . '</Currency>
							<Language>' . $this->constants->language . '</Language>
							<Version>' . $this->constants->api_ver . '</Version>
						</Authority>
						<BookingId>' . $bookingId . '</BookingId>
						<CommitLevel>confirm</CommitLevel>
					  </xiRequest>
					</BookingCancel>
				  </soap:Body>
				</soap:Envelope>';

		return $xml;
	}

	function convertDate( $date ) {
		//converts a date from the custom format into standart
		$formatstring = $this->constants->input_date_format;
		$delimiter    = strpos( $date, "/" ) !== false ? "/" : "-";
		$x            = explode( $delimiter, $date );
		$y            = explode( $delimiter, $formatstring );

		$dpos = array_search( 'dd', $y );
		$day  = $x[ $dpos ];

		$mpos  = array_search( 'mm', $y );
		$month = $x[ $mpos ];

		$ypos = array_search( 'yyyy', $y );
		if ($ypos==false) {
			$ypos = array_search( 'yy', $y );
		}
		$year = $x[ $ypos ];

		return $year . "-" . $month . "-" . $day;
	}

	function getArrayFromXmlResponseAvail( $response ) {
		$availability = array();
		// GET XML RESPONSE
		$soap = simplexml_load_string( $response );
		$soap->registerXPathNamespace( "soap", "http://www.w3.org/2003/05/soap-envelope" );
		$data = $soap->xpath( '//soap:Body' );

		// CONVERT OBJECT TO ARRAY
		$array = json_decode( json_encode( (array) $data ), 1 );

		// GET AVAILABILITY FROM XML RESPONSE
		if ( isset( $array[0]['AvailabilitySearchResponse']['AvailabilitySearchResult']['HotelAvailability'] ) ) {
			$availability = $array[0]['AvailabilitySearchResponse']['AvailabilitySearchResult']['HotelAvailability'];
		}

		if ( array_key_exists('Hotel',$availability)) {
			$availability = array( '0' => $availability );
		}

		foreach ( $availability as &$availability_entry ) {
			$result = $availability_entry['Result'];
			if ( array_key_exists('@attributes',$result) ) {
				$availability_entry['Result'] = array( '0' => $result );
			}
			foreach ( $availability_entry['Result'] as &$result_entry ) {
				$room = $result_entry['Room'];

				if ( array_key_exists('Price',$room) ) {
					$result_entry['Room'] = array( '0' => $room );
				}
			}
		}

		return $availability;
	}

	function getArrayFromXmlResponseBooking( $response ) {
		// GET XML RESPONSE
		$soap = simplexml_load_string( $response );
		$soap->registerXPathNamespace( "soap", "http://www.w3.org/2003/05/soap-envelope" );
		$data = $soap->xpath( '//soap:Body' );
		// CONVERT OBJECT TO ARRAY
		$array = json_decode( json_encode( (array) $data ), 1 );

		// GET AVAILABILITY FROM XML RESPONSE
		$booking = array();
		if ( isset( $array[0]['BookingCreateResponse']['BookingCreateResult'] ) ) {
			$booking = $array[0]['BookingCreateResponse']['BookingCreateResult'];
		} else {
			return $array[0];
		}

		if ( array_key_exists('Room',$booking['Booking']['HotelBooking']) ) {
			$booking['Booking']['HotelBooking'] = array( '0' => $booking['Booking']['HotelBooking'] );
		}

		foreach ( $booking['Booking']['HotelBooking'] as &$hotelbooking_entry ) {

			$room_entry = &$hotelbooking_entry['Room'];

			if ( array_key_exists('SellingPrice', $room_entry['NightCost']) ) {

				$room_entry['NightCost'] = array( '0' => $room_entry['NightCost'] );

			}
			if ( array_key_exists('Message',$room_entry['Messages']) && array_key_exists('Type',$room_entry['Messages']['Message']) ) {
				$room_entry['Messages']['Message'] = array( '0' => $room_entry['Messages']['Message'] );
			}
			if ( array_key_exists('Amount',$room_entry['CanxFees']['Fee'])) {
				$room_entry['CanxFees']['Fee'] = array( '0' => $room_entry['CanxFees']['Fee'] );
			}
		}


		//to do: make the structure of the array independent of the no. of nights, messages and fees.

		return $booking;
	}

}